compositor_render/src/transformations/web_renderer/transformation.rs

```diff
@@ -0,0 +1,177 @@
+        &self,
+        params: &TransformationParams,
+        _sources: &HashMap<NodeId, Texture>,
+        target: &Texture,
+    ) -> Result<(), Box<dyn Error>> {
+        let TransformationParams::String(url) = params else {
+            return Err(Box::new(WebRendererApplyError::InvalidParam));
+        };
+
+        let mut session_ids = self.session_ids.borrow_mut();
+
+        let session_id = match session_ids.get(url) {
+            Some(id) => id,
+            None => {
+                let size = target.size();
+                let id = self.api.new_session(
+                    url,
+                    Resolution {
+                        width: size.width as usize,
+                        height: size.height as usize,
+                    },
+                )?;
+
+                session_ids.insert(url.to_owned(), id);
+                session_ids.get(url).unwrap()
+            }
+        };
+
+        let frame = self.api.get_frame(session_id)?;
+        if !frame.is_empty() {
+            self.upload(&frame, target)?;
+        }
+
+        Ok(())
+    }
+
+    fn registry_key(&self) -> TransformationRegistryKey {
+        TransformationRegistryKey("web_renderer".to_owned())
+    }
+}
+
+impl Drop for WebRenderer {
+    fn drop(&mut self) {
+        if let Err(err) = self.renderer_process.kill() {
+            warn!("Failed to stop web renderer process: {err}");
+        }
+    }
+}
+
+impl WebRenderer {
+    pub fn new(wgpu_ctx: Rc<WgpuCtx>, port: u16) -> Result<Self, WebRendererNewError> {
+        let api = ElectronApi::new(port);
+        let renderer_process = Self::init_web_renderer(port)?;
+        let session_ids = RefCell::new(HashMap::new());
+
+        Ok(Self {
+            api,
+            wgpu_ctx,
+            renderer_process,
+            session_ids,
+        })
+    }
+
+    fn init_web_renderer(port: u16) -> Result<process::Child, WebRendererNewError> {
+        let web_renderer_path = env::current_exe()
+            .map_err(WebRendererNewError::WebRendererNotFound)?
+            .parent()
+            .unwrap()
+            .join("../../web_renderer");
+
+        let install_exit_code = Command::new("npm")
+            .arg("install")
+            .current_dir(&web_renderer_path)
+            .status()
+            .map_err(WebRendererNewError::WebRendererInitError)?;
+        if !install_exit_code.success() {
+            return Err(WebRendererNewError::WebRendererInstallError);
+        }
+
+        let renderer_process = Command::new("npm")
+            .args(["run", "start", "--", "--", &port.to_string()])
+            .current_dir(web_renderer_path)
+            .spawn()
+            .map_err(WebRendererNewError::WebRendererInitError)?;
+
+        Ok(renderer_process)
+    }
+
+    fn upload(&self, data: &[u8], target: &Texture) -> Result<(), WebRendererApplyError> {
+        let size = target.size();
+        let img = image::load_from_memory(data)?;
+
+        self.wgpu_ctx.queue.write_texture(
```

Comment:
You could do a safety check whether the image has the correct resolution before this.